File: shared/src/main/java/com/runtracker/shared/data/repository/PersonalBestRepository.kt
package com.runtracker.shared.data.repository

import com.runtracker.shared.data.db.PersonalBestDao
import com.runtracker.shared.data.model.PersonalBest
import com.runtracker.shared.data.model.Run
import com.runtracker.shared.data.model.WorkoutType
import kotlinx.coroutines.flow.Flow

class PersonalBestRepository(
    private val personalBestDao: PersonalBestDao
) {
    fun getAllPersonalBestsFlow(): Flow<List<PersonalBest>> = personalBestDao.getAllFlow()
    
    suspend fun getAllPersonalBests(): List<PersonalBest> = personalBestDao.getAll()
    
    fun getPersonalBestFlow(distanceMeters: Int): Flow<PersonalBest?> = 
        personalBestDao.getByDistanceFlow(distanceMeters)
    
    suspend fun getPersonalBest(distanceMeters: Int): PersonalBest? = 
        personalBestDao.getByDistance(distanceMeters)
    
    suspend fun getPersonalBestForWorkout(distanceMeters: Int, workoutType: WorkoutType): PersonalBest? =
        personalBestDao.getByDistanceAndWorkoutType(distanceMeters, workoutType)
    
    /**
     * Check if a run sets any new personal bests and update accordingly
     * Returns list of new PBs achieved
     */
    suspend fun checkAndUpdatePersonalBests(run: Run): List<PersonalBestUpdate> {
        val updates = mutableListOf<PersonalBestUpdate>()
        
        // Check each standard distance that the run covers
        for (distance in PersonalBest.STANDARD_DISTANCES) {
            if (run.distanceMeters >= distance) {
                val timeForDistance = calculateTimeForDistance(run, distance)
                if (timeForDistance != null) {
                    val update = checkDistancePB(run, distance, timeForDistance)
                    if (update != null) {
                        updates.add(update)
                    }
                }
            }
        }
        
        return updates
    }
    
    private suspend fun checkDistancePB(run: Run, distanceMeters: Int, timeMillis: Long): PersonalBestUpdate? {
        val existingPB = personalBestDao.getByDistance(distanceMeters)
        val paceSecondsPerKm = (timeMillis / 1000.0) / (distanceMeters / 1000.0)
        
        var newFastestTime = false
        var newLowestHr = false
        
        val updatedPB = if (existingPB == null) {
            // First time for this distance
            newFastestTime = true
            if (run.avgHeartRate != null) newLowestHr = true
            
            PersonalBest(
                distanceMeters = distanceMeters,
                fastestTimeMillis = timeMillis,
                fastestTimeRunId = run.id,
                fastestTimeDate = run.startTime,
                fastestTimePaceSecondsPerKm = paceSecondsPerKm,
                lowestAvgHeartRate = run.avgHeartRate,
                lowestHrRunId = if (run.avgHeartRate != null) run.id else null,
                lowestHrDate = if (run.avgHeartRate != null) run.startTime else null,
                lowestHrTimeMillis = if (run.avgHeartRate != null) timeMillis else null
            )
        } else {
            var updated = existingPB
            
            // Check fastest time
            if (existingPB.fastestTimeMillis == null || timeMillis < existingPB.fastestTimeMillis) {
                newFastestTime = true
                updated = updated.copy(
                    fastestTimeMillis = timeMillis,
                    fastestTimeRunId = run.id,
                    fastestTimeDate = run.startTime,
                    fastestTimePaceSecondsPerKm = paceSecondsPerKm
                )
            }
            
            // Check lowest heart rate (only if HR data available)
            if (run.avgHeartRate != null) {
                if (existingPB.lowestAvgHeartRate == null || run.avgHeartRate < existingPB.lowestAvgHeartRate) {
                    newLowestHr = true
                    updated = updated.copy(
                        lowestAvgHeartRate = run.avgHeartRate,
                        lowestHrRunId = run.id,
                        lowestHrDate = run.startTime,
                        lowestHrTimeMillis = timeMillis
                    )
                }
            }
            
            updated
        }
        
        if (newFastestTime || newLowestHr) {
            if (existingPB == null) {
                personalBestDao.insert(updatedPB)
            } else {
                personalBestDao.update(updatedPB)
            }
            
            return PersonalBestUpdate(
                distanceMeters = distanceMeters,
                distanceName = PersonalBest.getDistanceName(distanceMeters),
                newFastestTime = newFastestTime,
                newLowestHr = newLowestHr,
                timeMillis = timeMillis,
                avgHeartRate = run.avgHeartRate,
                previousFastestTime = existingPB?.fastestTimeMillis,
                previousLowestHr = existingPB?.lowestAvgHeartRate
            )
        }
        
        return null
    }
    
    /**
     * Calculate the time it took to complete a specific distance within a run
     * Uses splits if available, otherwise estimates based on average pace
     */
    private fun calculateTimeForDistance(run: Run, distanceMeters: Int): Long? {
        val distanceKm = distanceMeters / 1000
        
        // If we have splits, use them for more accurate time
        if (run.splits.isNotEmpty() && run.splits.size >= distanceKm) {
            var totalTime = 0L
            for (i in 0 until distanceKm) {
                totalTime += run.splits[i].durationMillis
            }
            return totalTime
        }
        
        // Otherwise estimate based on average pace
        if (run.avgPaceSecondsPerKm > 0) {
            return (run.avgPaceSecondsPerKm * distanceKm * 1000).toLong()
        }
        
        return null
    }
    
    /**
     * Get the pace at each kilometer for a personal best run
     * Used for compete mode comparison
     */
    suspend fun getPersonalBestSplits(distanceMeters: Int, runDao: com.runtracker.shared.data.db.RunDao): List<Long>? {
        val pb = personalBestDao.getByDistance(distanceMeters) ?: return null
        val runId = pb.fastestTimeRunId ?: return null
        val run = runDao.getRunById(runId) ?: return null
        
        return run.splits.map { it.durationMillis }
    }
}

data class PersonalBestUpdate(
    val distanceMeters: Int,
    val distanceName: String,
    val newFastestTime: Boolean,
    val newLowestHr: Boolean,
    val timeMillis: Long,
    val avgHeartRate: Int?,
    val previousFastestTime: Long?,
    val previousLowestHr: Int?
) {
    val timeImprovement: Long? get() = previousFastestTime?.let { it - timeMillis }
    val hrImprovement: Int? get() = previousLowestHr?.let { it - (avgHeartRate ?: 0) }
    
    val timeImprovementFormatted: String? get() {
        val improvement = timeImprovement ?: return null
        val seconds = improvement / 1000
        val minutes = seconds / 60
        val secs = seconds % 60
        return if (minutes > 0) "-${minutes}m ${secs}s" else "-${secs}s"
    }
}

File: iosApp/GoSteady/Sources/Repositories/PersonalBestRepository.swift
import Foundation
import SwiftData

final class PersonalBestRepository {
    private let context: ModelContext

    init(context: ModelContext) {
        self.context = context
    }

    // MARK: - Fetch

    func fetchAll() -> [PersonalBest] {
        let descriptor = FetchDescriptor<PersonalBest>(sortBy: [SortDescriptor(\.distanceMeters)])
        return (try? context.fetch(descriptor)) ?? []
    }

    func fetchByDistance(_ distanceMeters: Int) -> PersonalBest? {
        let descriptor = FetchDescriptor<PersonalBest>(predicate: #Predicate { $0.distanceMeters == distanceMeters })
        return try? context.fetch(descriptor).first
    }

    func fetchById(_ id: UUID) -> PersonalBest? {
        let descriptor = FetchDescriptor<PersonalBest>(predicate: #Predicate { $0.id == id })
        return try? context.fetch(descriptor).first
    }

    // MARK: - Check and Update

    /// Check if a run sets any new personal bests and update accordingly.
    /// Returns a list of PB updates achieved.
    func checkAndUpdatePersonalBests(run: Run) -> [PersonalBestUpdate] {
        var updates: [PersonalBestUpdate] = []

        for distance in PersonalBest.standardDistances {
            guard run.distanceMeters >= Double(distance) else { continue }
            guard let timeForDistance = calculateTimeForDistance(run: run, distanceMeters: distance) else { continue }
            if let update = checkDistancePB(run: run, distanceMeters: distance, timeMillis: timeForDistance) {
                updates.append(update)
            }
        }
        return updates
    }

    // MARK: - Private Helpers

    private func checkDistancePB(run: Run, distanceMeters: Int, timeMillis: Int64) -> PersonalBestUpdate? {
        let existing = fetchByDistance(distanceMeters)
        let paceSecPerKm = (Double(timeMillis) / 1000.0) / (Double(distanceMeters) / 1000.0)

        var newFastest = false
        var newLowestHR = false

        if let pb = existing {
            // Check fastest time
            if pb.fastestTimeMillis == nil || timeMillis < (pb.fastestTimeMillis ?? Int64.max) {
                newFastest = true
                pb.fastestTimeMillis = timeMillis
                pb.fastestTimeRunId = run.id
                pb.fastestTimeDate = run.startTime
                pb.fastestTimePaceSecondsPerKm = paceSecPerKm
            }
            // Check lowest HR
            if let runHR = run.avgHeartRate {
                if pb.lowestAvgHeartRate == nil || runHR < (pb.lowestAvgHeartRate ?? Int.max) {
                    newLowestHR = true
                    pb.lowestAvgHeartRate = runHR
                    pb.lowestHrRunId = run.id
                    pb.lowestHrDate = run.startTime
                    pb.lowestHrTimeMillis = timeMillis
                }
            }
            if newFastest || newLowestHR {
                try? context.save()
            }
        } else {
            // First time
            newFastest = true
            if run.avgHeartRate != nil { newLowestHR = true }
            let pb = PersonalBest(
                distanceMeters: distanceMeters,
                fastestTimeMillis: timeMillis,
                fastestTimeRunId: run.id,
                fastestTimeDate: run.startTime,
                fastestTimePaceSecondsPerKm: paceSecPerKm,
                lowestAvgHeartRate: run.avgHeartRate,
                lowestHrRunId: run.avgHeartRate != nil ? run.id : nil,
                lowestHrDate: run.avgHeartRate != nil ? run.startTime : nil,
                lowestHrTimeMillis: run.avgHeartRate != nil ? timeMillis : nil
            )
            context.insert(pb)
            try? context.save()
        }

        guard newFastest || newLowestHR else { return nil }

        return PersonalBestUpdate(
            distanceMeters: distanceMeters,
            distanceName: PersonalBest.getDistanceName(distanceMeters),
            newFastestTime: newFastest,
            newLowestHr: newLowestHR,
            timeMillis: timeMillis,
            avgHeartRate: run.avgHeartRate,
            previousFastestTime: existing?.fastestTimeMillis,
            previousLowestHr: existing?.lowestAvgHeartRate
        )
    }

    private func calculateTimeForDistance(run: Run, distanceMeters: Int) -> Int64? {
        let distanceKm = distanceMeters / 1000

        // Use splits for accuracy if available
        if run.splits.count >= distanceKm {
            var totalTime: Int64 = 0
            for i in 0..<distanceKm {
                totalTime += run.splits[i].durationMillis
            }
            return totalTime
        }

        // Estimate from average pace
        if run.avgPaceSecondsPerKm > 0 {
            return Int64(run.avgPaceSecondsPerKm * Double(distanceKm) * 1000)
        }

        return nil
    }

    /// Get the split times for the PB run at a given distance.
    func personalBestSplits(distanceMeters: Int, runRepository: RunRepository) -> [Int64]? {
        guard let pb = fetchByDistance(distanceMeters),
              let runId = pb.fastestTimeRunId,
              let run = runRepository.fetchById(runId) else { return nil }
        return run.splits.map(\.durationMillis)
    }
}

// MARK: - Supporting Types

struct PersonalBestUpdate {
    let distanceMeters: Int
    let distanceName: String
    let newFastestTime: Bool
    let newLowestHr: Bool
    let timeMillis: Int64
    let avgHeartRate: Int?
    let previousFastestTime: Int64?
    let previousLowestHr: Int?

    var timeImprovement: Int64? {
        guard let prev = previousFastestTime else { return nil }
        return prev - timeMillis
    }

    var timeImprovementFormatted: String? {
        guard let improvement = timeImprovement else { return nil }
        let seconds = improvement / 1000
        let minutes = seconds / 60
        let secs = seconds % 60
        if minutes > 0 {
            return "-\(minutes)m \(secs)s"
        }
        return "-\(secs)s"
    }
}
